src/blockstorage/prune.cpp

```diff
@@ -0,0 +1,418 @@
+
+#include <random>
+
+extern CBlockTreeDB *pblocktree;
+extern CDatabaseAbstract *pblockdb;
+extern CChain chainActive;
+extern CTweak<uint64_t> pruneIntervalTweak;
+extern CCriticalSection cs_LastBlockFile;
+extern std::set<int> setDirtyFileInfo;
+extern std::multimap<CBlockIndex *, CBlockIndex *> mapBlocksUnlinked;
+
+bool fPruneWithMask = DEFAULT_PRUNE_WITH_MASK;
+bool fHavePruned = false;
+uint64_t nPruneTarget = 0;
+uint64_t nDBUsedSpace = 0;
+uint64_t pruneHashMask = 0;
+const uint64_t ONE_THRESHOLD_PERCENT = std::numeric_limits<uint64_t>::max() / 100;
+uint8_t hashMaskThreshold;
+std::atomic<uint64_t> normalized_threshold;
+extern CTweakRef<uint8_t> hashMaskThresholdTweak;
+
+/** Global flag to indicate we should check to see if there are
+ *  block/undo files that should be deleted.  Set on startup
+ *  or if we allocate more file space when we're in prune mode
+ */
+bool fCheckForPruning = false;
+bool fPruneMode = false;
+
+extern void RelayNewXUpdate(const uint64_t key, const uint64_t val);
+extern bool AbortNode(const std::string &strMessage, const std::string &userMessage = "");
+
+std::string hashMaskThresholdValidator(const uint8_t &value, uint8_t *item, bool validate)
+{
+    std::string strReturn = std::string();
+    if (validate)
+    {
+        if (value > hashMaskThreshold)
+        {
+            std::ostringstream ret;
+            ret << "Sorry, your current hashMaskThreshold (" << std::to_string(hashMaskThreshold)
+                << ") is smaller than your proposed new threshold (" << std::to_string(value)
+                << "). You can only lower this number, not raise it.";
+            strReturn = ret.str();
+        }
+        else if (value == hashMaskThreshold)
+        {
+            // do nothing
+        }
+    }
+    else
+    {
+        uint8_t newValue = *item;
+        if (!pblocktree->WriteHashMaskThreshold(newValue))
+        {
+            strReturn = "Error writing new threshold to disk";
+        }
+        normalized_threshold = (newValue * ONE_THRESHOLD_PERCENT);
+        RelayNewXUpdate(XVer::BU_PRUNE_THRESHOLD, normalized_threshold);
+    }
+    return strReturn;
+}
+
+/** Generate a random list of 32 ints between 8 and 255 that will be used as important pruning bits*/
+void GenerateRandomPruningHashMask()
+{
+    // only generate new bits if we dont already have some
+    uint64_t read_hashMask = 0;
+    if (pblocktree->ReadHashMask(read_hashMask))
+    {
+        pruneHashMask = read_hashMask;
+        return;
+    }
+    uint64_t seed = GetTime();
+    std::mt19937_64 rand(seed); // Standard mersenne_twister_engine seeded with rd()
+    uint64_t hashMask64(rand());
+    pruneHashMask = 0;
+    pruneHashMask = pruneHashMask | hashMask64;
+    pblocktree->WriteHashMask(pruneHashMask);
+    pblocktree->WriteFlag("hashmaskexists", true);
+}
+
+/** Get important pruning bits from a block hash and compare their value to our pruning threshold*/
+bool hashMaskCompare(uint256 _blockHash)
```

Comment:
Let's name this better, esp since its not static.  We want to communicate that its about pruning and what is the sense of the returned boolean.  How about:

isBlockPrunable()

Note that if you use this name, it would also swap the return -- that is true would mean prune, and false means keep.

Reply:
it isnt static, but it is only used in this one file. I will change the name and remove it from the header.